libportability-gfx/src/impls.rs

```diff
@@ -1676,83 +1674,95 @@ pub extern "C" fn gfxCreateGraphicsPipelines(
-                        mask_write: state.writeMask,
-                        op_fail: conv::map_stencil_op(state.failOp),
-                        op_depth_fail: conv::map_stencil_op(state.depthFailOp),
-                        op_pass: conv::map_stencil_op(state.passOp),
+                    fn map_stencil_state(state: VkStencilOpState) -> pso::StencilFace {
+                        // TODO: reference value
+                        pso::StencilFace {
+                            fun: conv::map_compare_op(state.compareOp),
+                            mask_read: state.compareMask,
+                            mask_write: state.writeMask,
+                            op_fail: conv::map_stencil_op(state.failOp),
+                            op_depth_fail: conv::map_stencil_op(state.depthFailOp),
+                            op_pass: conv::map_stencil_op(state.passOp),
+                        }
                     }
-                }
 
-                let stencil_test = if state.stencilTestEnable == VK_TRUE {
-                    pso::StencilTest::On {
-                        front: map_stencil_state(state.front),
-                        back: map_stencil_state(state.back),
+                    let stencil_test = if state.stencilTestEnable == VK_TRUE {
+                        pso::StencilTest::On {
+                            front: map_stencil_state(state.front),
+                            back: map_stencil_state(state.back),
+                        }
+                    } else {
+                        pso::StencilTest::Off
+                    };
+
+                    // TODO: depth bounds
+
+                    pso::DepthStencilDesc {
+                        depth: depth_test,
+                        depth_bounds: state.depthBoundsTestEnable == VK_TRUE,
+                        stencil: stencil_test,
                     }
-                } else {
-                    pso::StencilTest::Off
-                };
+                })
+            }
+        } else {
+            None
+        };
 
-                // TODO: depth bounds
+        let mut viewport = None;
+        let mut scissor = None;
+        let mut blend_color = None;
+        let mut depth_bounds = None;
 
-                pso::DepthStencilDesc {
-                    depth: depth_test,
-                    depth_bounds: state.depthBoundsTestEnable == VK_TRUE,
-                    stencil: stencil_test,
+        if !rasterizer_discard {
+            let vp_state = unsafe { &*info.pViewportState };
+            let dyn_states = unsafe { info.pDynamicState.as_ref()
+                .map(|s| slice::from_raw_parts(s.pDynamicStates, s.dynamicStateCount as _))
+            };
+
+            if let Some(ds) = dyn_states {
+                for dyn_state in ds.iter() {
+                    match dyn_state {
+                        VkDynamicState::VK_DYNAMIC_STATE_VIEWPORT => unsafe {
+                            viewport = vp_state.pViewports.as_ref()
```

Comment:
these should be the other way around: if a dynamic state is present, the corresponding value we pass in should be None

Reply:
Overall, I slightly prefer the previous style for it being more expression-focused/functional

Reply:
Oops 😅 the double negative keeps tripping me up

Yeah, that's fair. I updated my commit to switch it back – how does this look?